vendor/github.com/manicminer/hamilton/msgraph/groups.go

```diff
@@ -0,0 +1,451 @@
+// groupId is the object ID of the group.
+// ownerId is the object ID of the owning object.
+func (c *GroupsClient) GetOwner(ctx context.Context, groupId, ownerId string) (*string, int, error) {
+	resp, status, _, err := c.BaseClient.Get(ctx, GetHttpRequestInput{
+		ValidStatusCodes: []int{http.StatusOK},
+		Uri: Uri{
+			Entity:      fmt.Sprintf("/groups/%s/owners/%s/$ref", groupId, ownerId),
+			Params:      url.Values{"$select": []string{"id,url"}},
+			HasTenantId: true,
+		},
+	})
+	if err != nil {
+		return nil, status, fmt.Errorf("GroupsClient.BaseClient.Get(): %v", err)
+	}
+	defer resp.Body.Close()
+	respBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, status, fmt.Errorf("ioutil.ReadAll(): %v", err)
+	}
+	var data struct {
+		Context string `json:"@odata.context"`
+		Type    string `json:"@odata.type"`
+		Id      string `json:"id"`
+		Url     string `json:"url"`
+	}
+	if err := json.Unmarshal(respBody, &data); err != nil {
+		return nil, status, fmt.Errorf("json.Unmarshal(): %v", err)
+	}
+	return &data.Id, status, nil
+}
+
+// AddOwners adds a new owner to a Group.
+// First populate the Owners field of the Group using the AppendOwner method of the model, then call this method.
+func (c *GroupsClient) AddOwners(ctx context.Context, group *Group) (int, error) {
+	var status int
+	for _, owner := range *group.Owners {
+		// don't fail if an owner already exists
+		checkOwnerAlreadyExists := func(resp *http.Response, o *odata.OData) bool {
+			if resp.StatusCode == http.StatusBadRequest {
+				if o.Error != nil {
+					return o.Error.Match(odata.ErrorAddedObjectReferencesAlreadyExist)
+				}
+			}
+			return false
+		}
+
+		data := struct {
+			Owner string `json:"@odata.id"`
+		}{
+			Owner: owner,
+		}
+		body, err := json.Marshal(data)
+		if err != nil {
+			return status, fmt.Errorf("json.Marshal(): %v", err)
+		}
+		_, status, _, err = c.BaseClient.Post(ctx, PostHttpRequestInput{
+			Body:             body,
+			ValidStatusCodes: []int{http.StatusNoContent},
+			ValidStatusFunc:  checkOwnerAlreadyExists,
+			Uri: Uri{
+				Entity:      fmt.Sprintf("/groups/%s/owners/$ref", *group.ID),
+				HasTenantId: true,
+			},
+		})
+		if err != nil {
+			return status, fmt.Errorf("GroupsClient.BaseClient.Post(): %v", err)
+		}
+	}
+	return status, nil
+}
+
+// RemoveOwners removes owners from a Group.
+// groupId is the object ID of the group.
+// ownerIds is a *[]string containing object IDs of owners to remove.
+func (c *GroupsClient) RemoveOwners(ctx context.Context, id string, ownerIds *[]string) (int, error) {
+	var status int
+	for _, ownerId := range *ownerIds {
+		// check for ownership before attempting deletion
+		if _, status, err := c.GetOwner(ctx, id, ownerId); err != nil {
+			if status == http.StatusNotFound {
+				continue
+			}
+			return status, err
+		}
+
+		// despite the above check, sometimes owners are just gone
+		checkOwnerGone := func(resp *http.Response, o *odata.OData) bool {
```

Comment:
(as above) wouldn't this indicate a bug?

Reply:
Similarly sometimes the read operation on a group indicates the presence of an owner when it's actually already been removed, so you end up with a GET showing an owner, but DELETEing that owner gets you a 403.